Objects/rangeobject.c

```diff
@@ -971,18 +971,22 @@ longrangeiter_setstate(longrangeiterobject *r, PyObject *state)
         if (cmp > 0)
             state = r->len;
     }
-    PyObject *new_len = PyNumber_Subtract(r->len, state);
-    if (new_len == NULL)
-        return NULL;
-    Py_SETREF(r->len, new_len);
     PyObject *product = PyNumber_Multiply(state, r->step);
     if (product == NULL)
         return NULL;
     PyObject *new_start = PyNumber_Add(r->start, product);
     Py_DECREF(product);
     if (new_start == NULL)
         return NULL;
-    Py_SETREF(r->start, new_start);
+    PyObject *new_len = PyNumber_Subtract(r->len, state);
+    if (new_len == NULL) {
+        Py_DECREF(new_start);
+        return NULL;
+    }
+    PyObject *tmp = r->start;
+    r->start = new_start;
+    Py_SETREF(r->len, new_len);
+    Py_DECREF(tmp);
```

Comment:
I don't see a scenario where we can't just use
```
Py_SETREF(r->len, new_len);
Py_SETREF(r->start, new_start);
```
(which would be a little easier to follow).

Both `new_len` and `new_start` are definitely new references we own, so even if `state` was a borrowed reference to `r->len` everything would be all right.

I'll leave it up to you though.

Reply:
Py_SETREF can only help when you assign a single object attribute, or if the attributes are independent. Two sequential `Py_SETREF`s can be a sign of a bug.

We do not check the type of `state`. It can be an arbitrary Python object with methods `__lt__`, `__gt__`, `__mul__` and `__rsub__`.  Therefore we do not control the types of `new_len` and `new_start`. Therefore we do not control the types of `r->len` and `r->start`. They can have `__del__` methods which release the GIL after assignment in `Py_SETREF`. When the GIL is released, the iterator object can be used in other thread when it is in inconsistent state -- new `r->len` and old `r->start`.

It never occurs in normal code (`state` is always an exact int), but if you try hard, you perhaps can reproduce this.

Reply:
Aha, that's a scenario I hadn't considered. Maybe the next time I am asked in some kind of Q&A or interview "do you have any regrets" I should mention `__del__` methods.